---
title: "Sunbeam Final Report"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
  rmd: "final_report.Rmd"
output:
  html_document:
    highlight: tango
    number_sections: yes
    theme: default
    toc: yes
    toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(viridis)

knitr::opts_chunk$set(
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  fig.width = 6.5
)

# DEBUGGING
if (!exists("snakemake")) {
  setClass(
    "Snakemake", 
    representation(input="character", output="character", params="character"))  
  snakemake <- new("Snakemake", input=c("/home/louistaylor/THING2/04_LowComplexitySeqs/20180501_SunbeamDatasetAnalysis/PLEASE_subset/sunbeam_output"))
}
taxa_ranks <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

```

```{r functions, include=FALSE}
MostSpecificTaxa <- function(agg, end.rank, ranks=taxa_ranks, label=FALSE, sep=":") {
  
  tax_climber <- function(otus, taxonomy, end, label, sep, ranks) {
    end_idx <- which(ranks==end)
    if (is.na(end_idx))
      stop("End rank not found in ranks")
    otus <- as.character(otus)
    if (all(is.na(otus)))
      return(rep(NA, length(otus)))
    taxa <- taxonomy[otus, 1:end_idx]
    min.ranks <- colnames(taxa)[apply(taxa, 1, function(x) max(which(!is.na(x))))]
    lowest <- taxa[cbind(otus, min.ranks)]
    if (label)
      paste(tolower(substr(min.ranks, 1, 1)), lowest, sep=sep)
    else
      lowest
  }
  
  .agg <- filter(agg, !is.na(otu))
  md <- .agg[,c(ranks, "otu")]
  md <- data.frame(distinct(md))
  rownames(md) <- md$otu
  minrank <- tax_climber(md$otu, md, end=end.rank, ranks=ranks, label=label, sep=sep)
  left_join(select(agg, otu), data.frame(otu=md$otu, MinRank=minrank))$MinRank
}

ParseOTUTable <- function(f) {
  # Adapted from qiimer::read_qiime_otu_table
  f <- file(f, 'rt')
  readLines(f, n=1)
  header <- readLines(f, n=1)
  col_names <- strsplit(header, "\t")[[1]]
  col_classes <- rep("numeric", times=length(col_names))
  col_classes[1] <- "character"
  col_classes[length(col_classes)] <- "character"
  otu_table <- read.table(
    f, col.names=col_names, colClasses = col_classes, as.is=TRUE, header=FALSE, 
    sep="\t", quote="")
  close(f)
  data_cols <- 2:(length(col_names)-1)
  sample_ids <- col_names[data_cols]
  otu_ids <- as.character(otu_table[,1])
  counts <- as.matrix(otu_table[, data_cols])
  rownames(counts) <- otu_ids
  metadata <- as.character(otu_table[, length(col_names)])
  names(metadata) <- otu_ids
  
  metadata <- data.frame(otu=otu_ids, taxonomy=metadata) %>% 
    # Separate taxonomy string into canonical ranks
    separate(taxonomy, into=taxa_ranks, sep="; ", extra="drop", fill="right") %>%
    # Convert missing assignments into explicit NAs
    mutate_at(vars(Kingdom:Species), funs(sub("[kpcofgs]__", "", .))) %>%
    mutate_at(vars(Kingdom:Species), funs(ifelse(.=="", NA, .))) %>%
    # Create binomial species names
    mutate(Species = ifelse(
        !is.na(Genus) & !is.na(Species),
        paste(as.character(Genus), as.character(Species)),
        Species))
  
  agg <- as.data.frame(counts[rowSums(counts) > 0, ]) %>%
    mutate(otu = rownames(.)) %>%
    gather(SampleID, count, -otu) %>%
    mutate_at(vars(SampleID, otu), "as.factor") %>%
    left_join(metadata) %>%
    mutate(
      MinRankSpecies = MostSpecificTaxa(., "Species"),
      MinRankPhylum = MostSpecificTaxa(., "Phylum"))
  agg
}

MakeHeatmapData <- function(agg, yaxis.rank="Genus", min.samples=1, 
                            hclust_method="average", ...) {
  
  .agg <- agg %>%
    mutate(MinRank = MostSpecificTaxa(agg, end.rank=yaxis.rank, ...)) %>%
    # Aggregate counts by minimum available rank
    group_by(SampleID, Kingdom, MinRank) %>%
    summarize(count = sum(count)) %>%
    # Update proportions
    group_by(SampleID) %>%
    mutate(proportion=count/sum(count))

  # Convert to matrix form
  .mat <- reshape2::dcast(
    .agg, MinRank ~ SampleID, value.var="proportion", fill = 0) %>%
    filter(!is.na(MinRank))
  rownames(.mat) <- .mat$MinRank
  .mat$MinRank <- NULL

  # Cluster and pull out row order
  minrank_order <- (dist(.mat, method="euclidean") %>%
                      hclust(method=hclust_method))$order

  .agg %>%
    # Reorder by clustering order
    mutate(MinRank = factor(MinRank, levels=rownames(.mat)[minrank_order])) %>%
    # Complete missing cases: fills missing combinations in with NA
    # instead of just omitting the row entirely (necessary to show blank cells)
    ungroup() %>%
    tidyr::complete(SampleID, nesting(MinRank)) %>%
    select(SampleID, Kingdom, MinRank, count, proportion) %>%
    distinct() %>%
    # Filter taxa that appear in fewer than required number of samples
    group_by(MinRank) %>%
    filter(sum(proportion > 0, na.rm=TRUE) > min.samples) %>%
    ungroup()
}

#' Uses the number of columns and number of rows in a heatmap to fix the
#' aspect ratio such that the cells are perfectly square.
#'
#' @param p a ggplot object with discrete x and y aesthetics set
#' @param fudge a fudge factor multiplying the aspect ratio (< 1 = wider,
#'   > 1 = taller). This does not appear to work right now.
#' @return the same object, with the aspect ratio fixed to be the (number of
#' rows)/(number of columns)
#' @export
make_square <- function(p, p_dims, fudge=1) {
  p + ggplot2::theme(aspect.ratio = (p_dims$nrows / p_dims$ncols) * fudge)
}

#' Filter taxon for a ggplot2 heatmap.
#' @return The data frame with taxa removed.
filter_taxon <- function(hmapData, threshold = 100){
  hmapData %>% 
    group_by(MinRank) %>% 
    filter(max(count) >= threshold) %>%
    ungroup() %>% 
    mutate(MinRank = fct_drop(MinRank))
}
```


# Quality Control Results
```{r qc-files}
decontam_report <- file.path(snakemake@input[['decontam']])
fastqc_report <- file.path(snakemake@input[['fastqc']])
n_samples <- 0
```

## Read quality

```{r fastqc-prep, fig.height=3, results='asis'}
.rexp <- "^(.*)_([12])$"
quality <- read.delim(fastqc_report, sep='\t') %>%
  reshape2::melt(id.vars="Samples", variable.name="Position", value.name = "Quality") %>%
  mutate(
    Position = sub("X", "", Position),
    Position = sub("\\.\\d+", "", Position, perl = TRUE),
    Position = as.numeric(Position)) %>%
  mutate(Sample=sub(.rexp,"\\1",Samples), Direction=sub(.rexp,"\\2",Samples)) %>%
  mutate(Direction = factor(Direction))

ave_quality <- group_by(quality, Direction, Position) %>%
  summarise(MeanQual = mean(Quality), SdQual = sd(Quality),
            SeQual = sd(Quality)/sqrt(n()), n = n()) %>% # chunyu: should be se instead sd
  mutate(LowQual = MeanQual - SdQual, HighQual = MeanQual + SdQual)

ggplot(ave_quality, aes(x=Position, y=MeanQual)) + 
  geom_errorbar(aes(ymin=LowQual, ymax=HighQual), color="grey60") +
  facet_wrap(~ Direction) +
  geom_line() +
  geom_point(shape=21) +
  theme_classic() + 
  labs(
    title="Sequence quality along read position",
    x='Position in sequence read', 
    y='Average quality score per sample')
```


## Read counts

```{r decontam-prep}
decontam_results <- read.delim(decontam_report) %>%
  mutate(Samples = factor(sub(".json", "", Samples)))
n_samples <- length(levels(decontam_results$Samples))
```

```{r decontam-readcounts, fig.height=1.5+0.1*n_samples, fig.width=5}
ggplot(decontam_results, aes(x=input, y=Samples)) +
  geom_segment(aes(yend=Samples, x=0, xend=input), color="grey30") +
  geom_point() +
  scale_x_continuous(expand=c(0,0), limits=c(0, max(decontam_results$input)*1.1)) +
  theme_classic() +
  labs(
    y="Sample",
    x="Read Pairs",
    title="Read pairs per sample, before QC"
  ) 
```

## Read decontamination


```{r decontam-results, results='asis', fig.height=1.5+0.1*n_samples}
cat("Reads removed because they matched one of the pre-specified ")
cat("host/contaminant genomes are grouped into the 'Host/Contaminant' category. ")
cat("Reads removed due to low complexity or quality are grouped into the 'Low Quality' category.")
preprocess <- decontam_results %>%
  mutate(low_quality = (fwd_only + rev_only + dropped) / input) %>% 
  mutate(contaminant = host / input) %>%
  mutate(noncontaminant = nonhost / input) %>%
  select(Samples, low_quality, contaminant, noncontaminant) %>%
  tidyr::gather("metric", "value", low_quality:noncontaminant)

ggplot(preprocess, aes(y=value, x=Samples, fill=metric)) +
  geom_bar(stat="identity", position="fill", color="white") +
  coord_flip() +
  scale_y_continuous(expand=c(0,0), labels=scales::percent) +
  scale_fill_manual("",
    labels=c("Host", "Low Quality", "Retained"),
    values=c("grey60", "grey80", "#029ae5ff")) +
  theme_classic() +
  labs(
    title="Host/low-quality read percentage",
    x="Sample", 
    y="Percentage of reads")
```

# Taxonomy Results

```{r taxa-files}
classify_report <- file.path(snakemake@input[['classify']])
```

```{r taxa-prep}

agg <- ParseOTUTable(classify_report) %>%
  mutate(SampleID = str_replace(SampleID, '^X', '')) %>%
  mutate(SampleID = str_replace(SampleID, '.taxa$', ''))
phyla.hmap.dat <- MakeHeatmapData(agg, yaxis="Phylum", min.samples = max(1, n_samples*0.2)) %>%
  filter_taxon()
phyla.hmap <- phyla.hmap.dat %>%
  ggplot(aes(SampleID, MinRank, fill=log10(count))) +
  geom_tile(color="grey80", size=0.4) +
  viridis::scale_fill_viridis(na.value="white", option = "A", labels=function(x) 10^(as.numeric(x))) +
  theme_classic() +
  theme(
    axis.line = element_blank(),
    panel.border = element_blank(),
    axis.text.x = element_text(angle=90, hjust=1, vjust=0.5),
    axis.title.x = element_blank()
  ) +
  labs(
    title="Phylum (or higher) level abundances",
    y="Resolved taxonomic rank",
    fill="Read counts"
  )
phyla.dims <- list(
  ncols = length(unique(phyla.hmap.dat$SampleID)),
  nrows = length(unique(phyla.hmap.dat$MinRank)))

family.hmap.dat <- MakeHeatmapData(agg, yaxis="Family", min.samples = max(1, n_samples*0.2)) %>%
  filter_taxon()
family.hmap <- family.hmap.dat %>%
  ggplot(aes(SampleID, MinRank, fill=log10(count))) +
  geom_tile(color="grey80", size=0.4) +
  viridis::scale_fill_viridis(na.value="white", option = "A", labels=function(x) 10^(as.numeric(x))) +
  theme_classic() +
  theme(
    axis.line = element_blank(),
    panel.border = element_blank(),
    axis.text.x = element_text(angle=90, hjust=1, vjust=0.5),
    axis.title.x = element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) +
  labs(
    title="Family (or higher) level abundances",
    y="Resolved taxonomic rank",
    fill="Read counts"
  )
family.dims <- list(
  ncols = length(unique(family.hmap.dat$SampleID)),
  nrows = length(unique(family.hmap.dat$MinRank)))

genera.hmap.dat <- MakeHeatmapData(agg, yaxis="Genus", min.samples = max(1, n_samples*0.2)) %>%
  filter_taxon(50)
genera.hmap <- genera.hmap.dat %>%
  ggplot(aes(SampleID, MinRank, fill=log10(count))) +
  geom_tile(color="grey80", size=0.4) +
  viridis::scale_fill_viridis(na.value="white", option = "A", labels=function(x) 10^(as.numeric(x))) +
  theme_classic() +
  theme(
    axis.line = element_blank(),
    panel.border = element_blank(),
    axis.text.x = element_text(angle=90, hjust=1, vjust=0.5),
    axis.title.x = element_blank()
  ) +
  labs(
    title="Genus (or higher) level abundances",
    y="Resolved taxonomic rank",
    fill="Read counts"
  )
genera.dims <- list(
  ncols = length(unique(genera.hmap.dat$SampleID)),
  nrows = length(unique(genera.hmap.dat$MinRank)))

species.hmap.dat <- MakeHeatmapData(agg, yaxis="Species", min.samples=0) %>%
  filter_taxon(50)
species.hmap <- species.hmap.dat %>%
  ggplot(aes(SampleID, MinRank, fill=log10(count))) +
  geom_tile(color="grey80", size=0.4) +
  viridis::scale_fill_viridis(na.value="white", option = "A", labels=function(x) 10^(as.numeric(x))) +
  theme_classic() +
  theme(
    axis.line = element_blank(),
    panel.border = element_blank(),
    axis.text.x = element_text(angle=90, hjust=1, vjust=0.5),
    axis.title.x = element_blank()
  ) +
  labs(
    title="Species (or higher) level abundances",
    y="Resolved taxonomic rank",
    fill="Read counts"
  )
species.dims <- list(
  ncols = length(unique(species.hmap.dat$SampleID)),
  nrows = length(unique(species.hmap.dat$MinRank)))

```

## Phylum-level heatmaps


Showing taxa at the phyla level or higher that appear in at least `r max(1, n_samples*0.2)` samples.

```{r phyla-heatmap, fig.height=1.5+0.1*phyla.dims$nrows, fig.width=2.5+0.2*phyla.dims$ncols, results='asis'}
save.image(file="debug.RData")
if (nrow(phyla.hmap.dat) > 0) {
  plot(make_square(phyla.hmap, phyla.dims))
} else {
  cat("**No phyla present in requisite number of samples**")
}

```

## Family-level heatmap

Showing taxa at the family level or higher that appear in at least `r max(1, n_samples*0.2)` samples.

```{r family-heatmap, fig.height=1.5+0.1*family.dims$nrows, fig.width=2.5+0.2*family.dims$ncols, results='asis'}
save.image(file="debug.RData")
if (nrow(family.hmap.dat) > 0) {
  plot(make_square(family.hmap, family.dims))
} else {
  cat("**No phyla present in requisite number of samples**")
}
```

## Genus-level heatmap

Showing taxa at the genus level or higher that appear in at least `r max(1, n_samples*0.2)` samples.

```{r genera-heatmap, fig.height=1.5+0.1*genera.dims$nrows, fig.width=2.5+0.2*genera.dims$ncols, results='asis'}
if (nrow(genera.hmap.dat) > 0) {
  plot(make_square(genera.hmap, genera.dims))
} else {
  cat("**No genera present in requisite number of samples**")
}
```

## Species-level heatmap

Showing all species level annotations for all samples. Warning: this plot can be quite large.

```{r species-heatmap, fig.height=4+0.1*species.dims$nrows, fig.width=4.5+0.2*species.dims$ncols, results='asis'}
if (nrow(species.hmap.dat) > 0) {
  plot(make_square(species.hmap, species.dims))
} else {
  cat("**No species present in samples**")
}
```
